File: plot/src/main/java/net/fwitz/math/plot/binary/dual/color/DomainColoringDual.java
package net.fwitz.math.plot.binary.dual.color;

import net.fwitz.math.binary.dual.DualNumber;

import java.awt.*;

import static java.lang.Math.abs;
import static java.lang.Math.exp;
import static java.lang.Math.floor;
import static java.lang.Math.sin;
import static java.lang.Math.sqrt;
import static net.fwitz.math.binary.dual.DualNumber.dualNumber;

public class DomainColoringDual implements DualNumberColorFunction {
    private static final double TWO_PI = Math.PI * 2.0;

    public Color apply(DualNumber c, DualNumber z) {
        return complexHSB(z);
    }

    private static Color complexHSB(DualNumber z) {
        if (z.isNaN()) {
            return Color.DARK_GRAY;
        }
        if (z.isInfinite()) {
            return Color.WHITE;
        }
        if (z.x() == 0) {
            return (z.y() == 0) ? Color.BLACK : Color.LIGHT_GRAY;
        }

        double h = hue(z);
        double s = abs(sin(TWO_PI * z.abs()));
        double b = sqrt(sqrt(abs(sin(TWO_PI * z.y()) * sin(TWO_PI * z.x()))));
        double oneMsMb = 1.0 - s - b;
        double b2 = 0.5 * ((1 - s) + b + sqrt(oneMsMb * oneMsMb + 0.01));
        return hsb(h, sqrt(s), b2);
    }

    // Hrmmm... how to smash  [-Inf, 0, +Inf] into [0.0, 0.5, 1.0]...
    private static double hue(DualNumber z) {
        double hue = 0.5 / (1 + exp(-z.arg()));
        hue = (z.x() < 0) ? (hue + 0.25) : (hue + 0.75);
        return hue - floor(hue);
    }

    private static Color hsb(double h, double s, double b) {
        return Color.getHSBColor(f(h), f(s), f(b));
    }

    private static float f(double x) {
        final float f = (float) x;
        return (f < 0.0f) ? 0.0f : ((f > 1.0f) ? 1.0f : f);
    }

    public static void main(String[] args) {
        debug(5, 0.01);
        debug(4, 3);
        debug(3, 4);
        debug(0, 5);
        debug(-3, 4);
        debug(-4, 3);
        debug(-5, 0);
        debug(-4, -3);
        debug(-3, -4);
        debug(0.01, -5);
        debug(3, -4);
        debug(4, -3);
    }

    private static void debug(double x, double y) {
        final DualNumber z = dualNumber(x, y);
        System.out.printf("z=(%10.7f, %10.7f)  abs=%10.7f  arg=%10.7f  hue=%10.7f  color=%s\n",
                x, y, z.abs(), z.arg(), hue(z), complexHSB(z));
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/binary/dual/color/DomainColoringDual.kt
package net.fwitz.math.plot.binary.dual.color

import net.fwitz.math.binary.dual.DualNumber
import java.awt.Color
import kotlin.math.*

object DomainColoringDual : DualNumberColorFunction {
    private const val TWO_PI = Math.PI * 2.0

    private fun complexHSB(z: DualNumber): Color = when {
        z.isNaN -> Color.DARK_GRAY
        z.isInfinite -> Color.WHITE
        z.x == 0.0 -> if (z.y == 0.0) Color.BLACK else Color.LIGHT_GRAY
        else -> {
            val h = hue(z)
            val s = abs(sin(TWO_PI * z.abs))
            val b = sqrt(sqrt(abs(sin(TWO_PI * z.y) * sin(TWO_PI * z.x))))
            val oneMsMb = 1.0 - s - b
            val b2 = 0.5 * (1 - s + b + sqrt(oneMsMb * oneMsMb + 0.01))
            hsb(h, sqrt(s), b2)
        }
    }

    // Hrmmm... how to smash [-Inf, 0, +Inf] into [0.0, 0.5, 1.0]...
    private fun hue(z: DualNumber): Double {
        var hue = 0.5 / (1 + exp(-z.arg))
        hue = if (z.x < 0) hue + 0.25 else hue + 0.75
        return hue - floor(hue)
    }

    private fun hsb(h: Double, s: Double, b: Double) = Color.getHSBColor(f(h), f(s), f(b))

    private fun f(x: Double): Float {
        val f = x.toFloat()
        return when {
            f < 0.0f -> 0.0f
            f > 1.0f -> 1.0f
            else -> f
        }
    }

    @JvmStatic
    fun main(args: Array<String>) {
        debug(5.0, 0.01)
        debug(4.0, 3.0)
        debug(3.0, 4.0)
        debug(0.0, 5.0)
        debug(-3.0, 4.0)
        debug(-4.0, 3.0)
        debug(-5.0, 0.0)
        debug(-4.0, -3.0)
        debug(-3.0, -4.0)
        debug(0.01, -5.0)
        debug(3.0, -4.0)
        debug(4.0, -3.0)
    }

    private fun debug(x: Double, y: Double) {
        val z = DualNumber(x, y)
        println(
            "z=(%10.7f, %10.7f)  abs=%10.7f  arg=%10.7f  hue=%10.7f  color=%s\n".format(
                x, y, z.abs, z.arg, hue(z), complexHSB(z)
            )
        )
    }

    override fun invoke(c: DualNumber, z: DualNumber) = complexHSB(z)
}